packages/react/src/utils/renderComponent.tsx

```diff
@@ -208,11 +208,23 @@ const renderComponent = <P extends {}>(
     : {}
 
   // Resolve styles using resolved variables, merge results, allow props.styles to override
-  const mergedStyles: ComponentSlotStylesPrepared = mergeComponentStyles(
-    theme.componentStyles[displayName],
-    withDebugId({ root: props.design }, 'props.design'),
-    withDebugId({ root: props.styles }, 'props.styles'),
-    withDebugId({ root: animationCSSProp }, 'props.animation'),
+  // TODO: update mergeComponentStyles to cache its results based on theme object and prop combinations.
```

Comment:
@levithomason @jdhuntington regarding this comment, I made updates (added `mergeComponentStylesWithCache`) for caching.. This required to do some changes in the factories.ts (in the way of how the defaultProps, userProps and overrideProps' styles are merged (I am considering them all to be object for now) in order for the hash function to work... 

Let me know what you think..

Reply:
Update - there are some perf gains with this:
```
Example with 1500 Menus, with different combinations of props, that repeatedly appear:

master avg: 5370.17
current avg: 4541.11

Chat with popover

master avg: 626.84
current avg: 596.07

ChatDuplicateMessages

master avg: 557.09 
current avg: 398.57
```

I had to add hash (id) on the theme - if it is not specified I am generating it as the JSON.stringify of the theme object (this is costly, is that's why for now I added hash on the different themes, but we can review it later). If it is useful I can move these changes to a separate PR and iterate on that.

Reply:
Hm, I would expect this to be much faster.  Are we sure it is working?

I added some comments on the cached function, have a look.

Reply:
Also, it would be great to see the results of `yarn perf` on master as a baseline in this PR and another table with the current commit.

Reply:
Here is a diff:
```
| Example                          | MASTER | CURRENT
|                                  | Avg    | Avg
| -------------------------------- | ------ | ------
| ProviderMergeThemes.perf.tsx     | 751.44 | 675.26
| DropdownManyItems.perf.tsx       | 527.11 | 484.49
| ChatWithPopover.perf.tsx         | 502.11 | 445.43
| ChatDuplicateMessages.perf.tsx   | 472.07 | 422.86
| CustomToolbar.perf.tsx           | 61.43  | 57.85
| SliderMinimal.perf.tsx           | 21.2   | 19.96
| ListCommon.perf.tsx              | 17.87  | 15.69
| EmbedMinimal.perf.tsx            | 17.29  | 15.68
| DropdownMinimal.perf.tsx         | 13.66  | 14.16
| AttachmentSlots.perf.tsx         | 11.98  | 10.02
| CarouselMinimal.perf.tsx         | 11.61  | 11.26
| SplitButtonMinimal.perf.tsx      | 10.27  | 8.88
| CheckboxMinimal.perf.tsx         | 8.68   | 7.43
| ProviderMinimal.perf.tsx         | 6.3    | 6.58
| LoaderMinimal.perf.tsx           | 5.97   | 6.02
| ButtonSlots.perf.tsx             | 5.18   | 5
| ReactionMinimal.perf.tsx         | 4.94   | 4.8
| InputMinimal.perf.tsx            | 4.6    | 4.15
| ButtonMinimal.perf.tsx           | 4.07   | 3.35
| AttachmentMinimal.perf.tsx       | 3.99   | 4.33
| TextAreaMinimal.perf.tsx         | 3.59   | 3.43
| DialogMinimal.perf.tsx           | 2.7    | 2.48
| AvatarMinimal.perf.tsx           | 2.5    | 2.27
| AlertMinimal.perf.tsx            | 2.37   | 2.3
| MenuMinimal.perf.tsx             | 2.3    | 2.16
| ItemLayoutMinimal.perf.tsx       | 2.12   | 1.97
| MenuButtonMinimal.perf.tsx       | 1.99   | 1.76
| LabelMinimal.perf.tsx            | 1.94   | 1.63
| AccordionMinimal.perf.tsx        | 1.83   | 1.62
| TreeMinimal.perf.tsx             | 1.69   | 1.37
| HeaderSlots.perf.tsx             | 1.63   | 1.63
| SegmentMinimal.perf.tsx          | 1.37   | 1.32
| VideoMinimal.perf.tsx            | 1.35   | 1.32
| HierarchicalTreeMinimal.perf.tsx | 1.35   | 1.39
| ChatMinimal.perf.tsx             | 1.29   | 1.05
| FormMinimal.perf.tsx             | 1.23   | 1.17
| ToolbarMinimal.perf.tsx          | 1.23   | 1.46
| DividerMinimal.perf.tsx          | 1.18   | 1.18
| TableMinimal.perf.tsx            | 1.05   | 0.97
| GridMinimal.perf.tsx             | 1.03   | 1.01
| IconMinimal.perf.tsx             | 0.95   | 0.97
| RadioGroupMinimal.perf.tsx       | 0.94   | 0.9
| StatusMinimal.perf.tsx           | 0.87   | 0.73
| PopupMinimal.perf.tsx            | 0.85   | 0.92
| ListMinimal.perf.tsx             | 0.8    | 0.67
| ChatMinimal.perf.tsx             | 1.29   | 1.05
| FlexMinimal.perf.tsx             | 0.71   | 0.68
| HeaderMinimal.perf.tsx           | 0.69   | 0.59
| ImageMinimal.perf.tsx            | 0.64   | 0.5
| LayoutMinimal.perf.tsx           | 0.62   | 0.61
| AnimationMinimal.perf.tsx        | 0.5    | 0.48
| TextMinimal.perf.tsx             | 0.47   | 0.5
| PortalMinimal.perf.tsx           | 0.37   | 0.36
| RefMinimal.perf.tsx              | 0.23   | 0.19
| BoxMinimal.perf.tsx              | 0.21   | 0.19

```

Reply:
Please see comment: https://github.com/microsoft/fluent-ui-react/pull/2200#discussion_r364650144

Reply:
Cool, thanks for the diff.  It looks like this isn't making a substantial improvement.  Although many are a bit faster, some are actually slower too.

Let's see what happens with the immutable theme approach.  Flamegrill can also help show you where cycles are being spent.  You might want to compare approaches of hash/stringify with immutable and see where the time is going.

Reply:
I opened PR: https://github.com/microsoft/fluent-ui-react/pull/2226/files and reverted the changes in this PR. Left some comments, take a look when you can, and we can move the discussion there.